mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala

```diff
@@ -193,7 +193,8 @@ private[spark] class MatrixUDT extends UserDefinedType[Matrix] {
     }
   }
 
-  override def hashCode(): Int = 1994
+  // see [SPARK-8647], this achieves the needed constant hash code without constant no.
+  override def hashCode(): Int = this.getClass.getName.hashCode()
```

Comment:
@srowen @MechCoder Please see the jira https://issues.apache.org/jira/browse/SPARK-8647  for more details.